tests/functional/setups/manu_thermoporomech_nofrac_2d.py

```diff
@@ -0,0 +1,1219 @@
+        flux_ad = self.energy_flux([sd])
+        approx_energy_flux = flux_ad.value(self.equation_system)
+        error_energy_flux = ConvergenceAnalysis.l2_error(
+            grid=sd,
+            true_array=exact_energy_flux,
+            approx_array=approx_energy_flux,
+            is_scalar=True,
+            is_cc=False,
+            relative=True,
+        )
+
+        exact_force = self.exact_sol.poroelastic_force(sd=sd, time=t)
+        force_ad = self.stress([sd])
+        approx_force = force_ad.value(self.equation_system)
+        error_force = ConvergenceAnalysis.l2_error(
+            grid=sd,
+            true_array=exact_force,
+            approx_array=approx_force,
+            is_scalar=False,
+            is_cc=False,
+            relative=True,
+        )
+
+        # Store collected data in data class
+        collected_data = ManuThermoPoroMechSaveData(
+            approx_displacement=approx_displacement,
+            approx_darcy_flux=approx_darcy_flux,
+            approx_energy_flux=approx_energy_flux,
+            approx_force=approx_force,
+            approx_pressure=approx_pressure,
+            approx_temperature=approx_temperature,
+            error_displacement=error_displacement,
+            error_darcy_flux=error_darcy_flux,
+            error_energy_flux=error_energy_flux,
+            error_force=error_force,
+            error_pressure=error_pressure,
+            exact_displacement=exact_displacement,
+            error_temperature=error_temperature,
+            exact_darcy_flux=exact_darcy_flux,
+            exact_energy_flux=exact_energy_flux,
+            exact_force=exact_force,
+            exact_pressure=exact_pressure,
+            exact_temperature=exact_temperature,
+            time=t,
+        )
+
+        return collected_data
+
+
+class ManuThermoPoroMechExactSolution2d:
+    """Class containing the exact manufactured solution for the verification setup.
+
+    The exact solutions for the primary variables pressure, displacement, temperature,
+    are defined below, as well as the exact solutions for the secondary variables Darcy
+    flux and poroelastic force.
+
+    A heterogeneity is introduced in the permeability and Lamé parameters, so that these
+    take different values in the region x > 0.5 and y > 0.5. The primary variables
+    pressure and displacement are scaled with the heterogeneity, but in a reciprocal
+    way, so that the derived fluxes and stresses are continuous across the interface
+    between the regions with different material parameters. Moreover, the primary
+    variables are constructed to be zero at the interface, so that they are continuous
+    across the interface. This is necessary to ensure that the exact solution can be
+    used to verify convergence of the numerical solution.
+
+    The temperature variable was not scaled with the heterogeneity, since this would be
+    difficult to make work with the fluid conductivity (thus effective heterogeneity)
```

Comment:
I don't get the parenthesis.
```suggestion
    difficult to make work with the fluid heat conductivity (thus effective heterogeneity)
```